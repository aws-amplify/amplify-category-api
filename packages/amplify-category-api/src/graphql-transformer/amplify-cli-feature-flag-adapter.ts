import { FeatureFlags } from 'amplify-cli-core';
import { FeatureFlagProvider } from 'graphql-transformer-core';
import { FeatureFlagProvider as NewTransformerFFProvider } from '@aws-amplify/graphql-transformer-interfaces'

export class AmplifyCLIFeatureFlagAdapterBase implements FeatureFlagProvider {
  getBoolean(featureName: string, defaultValue?: boolean): boolean {
    return this.getValue<boolean>(featureName, 'boolean', defaultValue);
  }
  getNumber(featureName: string, defaultValue?: number): number {
    return this.getValue<number>(featureName, 'number', defaultValue);
  }
  getObject(): object {
    // Todo: for future extensibility
    throw new Error('Not implemented');
  }

  protected getValue<T extends number | boolean>(featureName: string, type: 'boolean' | 'number' | 'string', defaultValue: T): T {
    const keyName = `graphQLTransformer.${featureName}`;
    try {
      switch (type) {
        case 'boolean':
          return FeatureFlags.getBoolean(keyName) as T;
        case 'number':
          return FeatureFlags.getNumber(keyName) as T;
      }
    } catch (e) {
      if (defaultValue) {
        return defaultValue;
      }
      throw e;
    }
  }
}

// Mapping to new type to ensure the provider interface is implemented
export class AmplifyCLIFeatureFlagAdapter
 extends AmplifyCLIFeatureFlagAdapterBase implements NewTransformerFFProvider {}