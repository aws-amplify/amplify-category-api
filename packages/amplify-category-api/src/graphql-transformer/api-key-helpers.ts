import { $TSContext, CloudformationProviderFacade } from '@aws-amplify/amplify-cli-core';

/**
 *
 * @param context
 */
export async function hasApiKey(context: $TSContext): Promise<boolean> {
  const apiKeyConfig = await CloudformationProviderFacade.getApiKeyConfig(context);
  return !!apiKeyConfig && !!apiKeyConfig?.apiKeyExpirationDays;
}
