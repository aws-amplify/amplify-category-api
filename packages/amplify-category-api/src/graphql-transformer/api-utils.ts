import { ensureEnvParamManager } from '@aws-amplify/amplify-environment-parameters';
import {
  $TSContext, $TSObject, AmplifyCategories, ApiCategoryFacade, getGraphQLTransformerOpenSearchProductionDocLink,
} from 'amplify-cli-core';
import { printer } from 'amplify-prompts';
import { ResourceConstants } from 'graphql-transformer-common';

export async function searchablePushChecks(context: $TSContext, map: $TSObject, apiName: string): Promise<void> {
  const searchableModelTypes = Object.keys(map).filter((type) => map[type].includes('searchable') && map[type].includes('model'));
  if (searchableModelTypes.length) {
    const apiParameterManager = (await ensureEnvParamManager()).instance.getResourceParamManager(AmplifyCategories.API, apiName);
    const getInstanceType = (instanceTypeParam: string) => apiParameterManager.getParam(instanceTypeParam);
    const instanceType = getInstanceType(ResourceConstants.PARAMETERS.OpenSearchInstanceType)
      ?? getInstanceType(ResourceConstants.PARAMETERS.ElasticsearchInstanceType)
      ?? 't2.small.elasticsearch';
    if (instanceType === 't2.small.elasticsearch' || instanceType === 't3.small.elasticsearch') {
      const version = await ApiCategoryFacade.getTransformerVersion(context);
      const docLink = getGraphQLTransformerOpenSearchProductionDocLink(version);
      printer.warn(
        `Your instance type for OpenSearch is ${instanceType}, you may experience performance issues or data loss. Consider reconfiguring with the instructions here ${docLink}`,
      );
    }
  }
}
