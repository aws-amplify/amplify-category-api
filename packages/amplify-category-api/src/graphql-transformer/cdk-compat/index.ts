export { TransformerRootStack } from './root-stack';
export { TransformerNestedStack } from './nested-stack';
export { TransformerStackSythesizer } from './stack-synthesizer';
