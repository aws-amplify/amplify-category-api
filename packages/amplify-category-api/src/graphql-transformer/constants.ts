export const DESTRUCTIVE_UPDATES_FLAG = 'allow-destructive-graphql-schema-updates';
export const PROVIDER_NAME = 'awscloudformation';
export const PARAMETERS_FILENAME = 'parameters.json';
export const ROOT_APPSYNC_S3_KEY = 'amplify-appsync-files';
export const SCHEMA_FILENAME = 'schema.graphql';
export const SCHEMA_DIR_NAME = 'schema';
