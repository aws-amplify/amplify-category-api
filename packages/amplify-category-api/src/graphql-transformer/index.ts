export { getDirectiveDefinitions } from './directive-definitions';
export { getTransformerVersion } from './transformer-version';
export { transformGraphQLSchema } from './transform-graphql-schema';
