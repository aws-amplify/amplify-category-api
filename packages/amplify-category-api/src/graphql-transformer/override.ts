import { CfnResource } from "aws-cdk-lib";
import * as fs from "fs-extra";
import * as vm from "vm2";
import * as path from "path";
import _ from "lodash";
import { pathManager, stateManager } from "@aws-amplify/amplify-cli-core";
import { StackManager } from '@aws-amplify/graphql-transformer-core';
import { AmplifyApiGraphQlResourceStackTemplate } from "@aws-amplify/graphql-transformer-interfaces";
import { ConstructResourceMeta } from "./types/types";
import { convertToAppsyncResourceObj, getStackMeta } from "./types/utils";
import { getAppSyncAPIName } from "../provider-utils/awscloudformation/utils/amplify-meta-utils";

/**
 *
 * @param stackManager
 * @param overrideDir
 */
export function applyFileBasedOverride(stackManager: StackManager, overrideDirPath?: string): AmplifyApiGraphQlResourceStackTemplate {
  const overrideDir = overrideDirPath ?? path.join(pathManager.getBackendDirPath(), 'api', getAppSyncAPIName());
  const overrideFilePath = path.join(overrideDir, "build", "override.js");
  if (!fs.existsSync(overrideFilePath)) {
    return {};
  }

  const stacks: string[] = [];
  const amplifyApiObj: any = {};
  stackManager.rootStack.node.findAll().forEach(node => {
    const resource = node as CfnResource;
    if (resource.cfnResourceType === "AWS::CloudFormation::Stack") {
      stacks.push(node.node.id.split(".")[0]);
    }
  });

  stackManager.rootStack.node.findAll().forEach(node => {
    const resource = node as CfnResource;
    let pathArr;
    if (node.node.id === "Resource") {
      pathArr = node.node.path.split("/").filter(key => key !== node.node.id);
    } else {
      pathArr = node.node.path.split("/");
    }
    let constructPathObj: ConstructResourceMeta;
    if (resource.cfnResourceType) {
      constructPathObj = getStackMeta(pathArr, node.node.id, stacks, resource);
      if (!_.isEmpty(constructPathObj.rootStack)) {
        // api scope
        const field = constructPathObj.rootStack!.stackType;
        const { resourceName } = constructPathObj;
        _.set(amplifyApiObj, [field, resourceName], resource);
      } else if (!_.isEmpty(constructPathObj.nestedStack)) {
        const fieldType = constructPathObj.nestedStack!.stackType;
        const fieldName = constructPathObj.nestedStack!.stackName;
        const { resourceName } = constructPathObj;
        if (constructPathObj.resourceType.includes("Resolver")) {
          _.set(amplifyApiObj, [fieldType, fieldName, "resolvers", resourceName], resource);
        } else if (constructPathObj.resourceType.includes("FunctionConfiguration")) {
          _.set(amplifyApiObj, [fieldType, fieldName, "appsyncFunctions", resourceName], resource);
        } else {
          _.set(amplifyApiObj, [fieldType, fieldName, resourceName], resource);
        }
      }
    }
  });

  const appsyncResourceObj = convertToAppsyncResourceObj(amplifyApiObj);
  const overrideCode: string = fs.readFileSync(overrideFilePath, "utf-8");
  const sandboxNode = new vm.NodeVM({
    console: "inherit",
    timeout: 5000,
    sandbox: {},
    require: {
      context: "sandbox",
      builtin: ["path"],
      external: true
    }
  });
  // Remove these when moving override up to amplify-category-api level
  const { envName } = stateManager.getLocalEnvInfo();
  const { projectName } = stateManager.getProjectConfig();
  const projectInfo = {
    envName,
    projectName
  };
  try {
    sandboxNode.run(overrideCode, overrideFilePath).override(appsyncResourceObj, projectInfo);
  } catch (err) {
    throw new InvalidOverrideError(err);
  }
  return appsyncResourceObj;
}

/**
 *
 */
export class InvalidOverrideError extends Error {
  details: string;
  resolution: string;
  constructor(error: Error) {
    super("Executing overrides failed.");
    this.name = "InvalidOverrideError";
    this.details = error.message;
    this.resolution = "There may be runtime errors in your overrides file. If so, fix the errors and try again.";
    if ((Error as any).captureStackTrace) {
      (Error as any).captureStackTrace(this, InvalidOverrideError);
    }
  }
}
