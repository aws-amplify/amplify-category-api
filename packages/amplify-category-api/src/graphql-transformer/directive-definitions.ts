import { getAppSyncServiceExtraDirectives } from '@aws-amplify/graphql-transformer-core';
import { $TSContext } from '@aws-amplify/amplify-cli-core';
import { print } from 'graphql';
import { TransformerPluginProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { getTransformerFactoryV1 } from './transformer-factory';
import { getTransformerVersion } from './transformer-version';
import { constructTransformerChain } from '@aws-amplify/graphql-transformer';
import { loadCustomTransformersV2 } from './transformer-options-v2';

/**
 * Return the set of directive definitions for the project, includes both appsync and amplify supported directives.
 * This will return the relevant set determined by whether or not the customer is using GQL transformer v1 or 2 in their project.
 */
export const getDirectiveDefinitions = async (context: $TSContext, resourceDir: string): Promise<string> => {
  const transformerVersion = await getTransformerVersion(context);
  const transformList =
    transformerVersion === 2 ? await getTransformListV2(resourceDir) : await getTransformerFactoryV1(context, resourceDir)(true);

  const transformDirectives = transformList
    .map((transform) => [transform.directive, ...transform.typeDefinitions].map((node) => print(node)).join('\n'))
    .join('\n');

  return [getAppSyncServiceExtraDirectives(), transformDirectives].join('\n');
};

/**
 * Get the list of v2 transformers, including custom transformers.
 * @param resourceDir directory to search for custom transformer config in.
 * @returns the list of transformers, including any defined custom transformers.
 */
const getTransformListV2 = async (resourceDir: string): Promise<TransformerPluginProvider[]> =>
  constructTransformerChain({
    customTransformers: await loadCustomTransformersV2(resourceDir),
  });
