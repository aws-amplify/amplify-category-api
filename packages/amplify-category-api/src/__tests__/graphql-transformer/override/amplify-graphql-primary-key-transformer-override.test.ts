import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform, StackManager, ConflictHandlerType, SyncConfig } from '@aws-amplify/graphql-transformer-core';
import * as path from 'path';
import { PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { stateManager } from '@aws-amplify/amplify-cli-core';
import { applyFileBasedOverride } from '../../../graphql-transformer/override';

jest.spyOn(stateManager, 'getLocalEnvInfo').mockReturnValue({ envName: 'testEnvName' });
jest.spyOn(stateManager, 'getProjectConfig').mockReturnValue({ projectName: 'testProjectName' });

test('it overrides expected resources', () => {
  const validSchema = `
    type Test @model {
      email: String! @primaryKey
    }
  `;
  const config: SyncConfig = {
    ConflictDetection: 'VERSION',
    ConflictHandler: ConflictHandlerType.AUTOMERGE,
  };

  const transformer = new GraphQLTransform({
    transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
    resolverConfig: {
      project: config,
    },
    overrideConfig: {
      applyOverride: (stackManager: StackManager) => {
        return applyFileBasedOverride(stackManager, path.join(__dirname, 'primary-key-overrides'))
      },
      overrideFlag: true,
    },
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks.Test.Resources.CreateTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.GetTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.ListTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.DeleteTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.UpdateTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.SyncTestResolver).toMatchSnapshot();
  expect(out.stacks.Test.Resources.TestDataSource).toMatchSnapshot();
});
