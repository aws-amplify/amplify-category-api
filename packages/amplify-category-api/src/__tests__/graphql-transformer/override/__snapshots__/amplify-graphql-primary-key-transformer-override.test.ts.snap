// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it overrides expected resources 1`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "createTest",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "MutationcreateTestinit0FunctionMutationcreateTestinit0FunctionAppSyncFunction82753130",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "MutationcreateTestpreAuth0FunctionMutationcreateTestpreAuth0FunctionAppSyncFunctionC0EB74A1",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "MutationCreateTestDataResolverFnMutationCreateTestDataResolverFnAppSyncFunction005BA3A1",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Mutation\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"createTest\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Mutation",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 2`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "getTest",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpreAuth0FunctionQuerygetTestpreAuth0FunctionAppSyncFunction7DE868FF",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QueryGetTestDataResolverFnQueryGetTestDataResolverFnAppSyncFunctionCB6458E5",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Query\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"getTest\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Query",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 3`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "listTests",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "QuerylistTestspreAuth0FunctionQuerylistTestspreAuth0FunctionAppSyncFunction3CA4A90C",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QueryListTestsDataResolverFnQueryListTestsDataResolverFnAppSyncFunction177A117F",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Query\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"listTests\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Query",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 4`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "deleteTest",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "MutationdeleteTestpreAuth0FunctionMutationdeleteTestpreAuth0FunctionAppSyncFunctionC68C5724",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "MutationDeleteTestDataResolverFnMutationDeleteTestDataResolverFnAppSyncFunction9230CF9A",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Mutation\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"deleteTest\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Mutation",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 5`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "updateTest",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "MutationupdateTestinit0FunctionMutationupdateTestinit0FunctionAppSyncFunctionC2375285",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "MutationcreateTestpreAuth0FunctionMutationcreateTestpreAuth0FunctionAppSyncFunctionC0EB74A1",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "MutationUpdateTestDataResolverFnMutationUpdateTestDataResolverFnAppSyncFunction161162F8",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Mutation\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"updateTest\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Mutation",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 6`] = `
Object {
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "FieldName": "syncTests",
    "Kind": "PIPELINE",
    "PipelineConfig": Object {
      "Functions": Array [
        Object {
          "Fn::GetAtt": Array [
            "QuerysyncTestspreAuth0FunctionQuerysyncTestspreAuth0FunctionAppSyncFunction6F326BAB",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerygetTestpostAuth0FunctionQuerygetTestpostAuth0FunctionAppSyncFunctionE9BFAF7A",
            "FunctionId",
          ],
        },
        Object {
          "Fn::GetAtt": Array [
            "QuerySyncTestsDataResolverFnQuerySyncTestsDataResolverFnAppSyncFunction95B4DA0E",
            "FunctionId",
          ],
        },
      ],
    },
    "RequestMappingTemplate": Object {
      "Fn::Join": Array [
        "",
        Array [
          "$util.qr($ctx.stash.put(\\"typeName\\", \\"Query\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"syncTests\\"))
$util.qr($ctx.stash.put(\\"conditions\\", []))
$util.qr($ctx.stash.put(\\"metadata\\", {}))
$util.qr($ctx.stash.metadata.put(\\"dataSourceType\\", \\"AMAZON_DYNAMODB\\"))
$util.qr($ctx.stash.metadata.put(\\"apiId\\", \\"",
          Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"connectionAttributes\\", {}))
$util.qr($ctx.stash.put(\\"tableName\\", \\"",
          Object {
            "Ref": "TestTable",
          },
          "\\"))
$util.qr($ctx.stash.put(\\"deltaSyncTableTtl\\", \\"25\\"))
$util.toJson({})",
        ],
      ],
    },
    "ResponseMappingTemplate": "$util.toJson($ctx.prev.result)",
    "TypeName": "Query",
  },
  "Type": "AWS::AppSync::Resolver",
}
`;

exports[`it overrides expected resources 7`] = `
Object {
  "DependsOn": Array [
    "TestIAMRoleB964910C",
  ],
  "Properties": Object {
    "ApiId": Object {
      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
    },
    "DynamoDBConfig": Object {
      "AwsRegion": Object {
        "Ref": "AWS::Region",
      },
      "DeltaSyncConfig": Object {
        "BaseTableTTL": "20",
        "DeltaSyncTableName": "TestTable",
        "DeltaSyncTableTTL": "25",
      },
      "TableName": Object {
        "Ref": "TestTable",
      },
      "Versioned": true,
    },
    "Name": "TestTable",
    "ServiceRoleArn": Object {
      "Fn::GetAtt": Array [
        "TestIAMRoleB964910C",
        "Arn",
      ],
    },
    "Type": "AMAZON_DYNAMODB",
  },
  "Type": "AWS::AppSync::DataSource",
}
`;
